pyresample/gradient/__init__.py

```diff
@@ -406,3 +440,194 @@ def _concatenate_chunks(chunks):
+    return wrapper
+
+
+@ensure_3d_data
+def gradient_resampler(data, source_area, target_area, method='bilinear'):
+    """Do the gradient search resampling."""
+    from pyproj.transformer import Transformer
+
+    try:
+        src_x, src_y = source_area.get_proj_coords()
+        transformer = Transformer.from_crs(target_area.crs, source_area.crs)
+    except AttributeError:
+        src_x, src_y = source_area
+        # TODO: this is bad. We don't want to perform gradient search in lat/lon space (poles, dateshift line)
+        transformer = Transformer.from_crs(target_area.crs, CRS(proj='longlat'))
+
+    # TODO: this will crash when the target area is a swath def.
+    dst_x, dst_y = transformer.transform(*target_area.get_proj_coords())
+    src_gradient_xl, src_gradient_xp = np.gradient(src_x, axis=[0, 1])
+    src_gradient_yl, src_gradient_yp = np.gradient(src_y, axis=[0, 1])
+
+    return _gradient_resample_data(data, src_x, src_y,
+                                   src_gradient_xl, src_gradient_xp,
+                                   src_gradient_yl, src_gradient_yp,
+                                   dst_x, dst_y,
+                                   method=method)
+
+
+def gradient_resampler_indices(source_area, target_area, method='bilinear', block_info=None, **kwargs):
+    """Do the gradient search resampling, returning the resulting indices."""
+    from pyproj.transformer import Transformer
+    try:
+        src_x, src_y = source_area.get_proj_coords()
+        transformer = Transformer.from_crs(target_area.crs, source_area.crs, always_xy=True)
+    except AttributeError:
+        src_x, src_y = np.asarray(source_area.lons), np.asarray(source_area.lats)
+        # TODO: this is bad. We don't want to perform gradient search in lat/lon space (poles, dateshift line)
+        transformer = Transformer.from_crs(target_area.crs, CRS(proj='longlat'))
+
+    # TODO: this will crash when the target area is a swath def.
+    dst_x, dst_y = transformer.transform(*target_area.get_proj_coords())
+    src_gradient_xl, src_gradient_xp = np.gradient(src_x, axis=[0, 1])
+    src_gradient_yl, src_gradient_yp = np.gradient(src_y, axis=[0, 1])
+
+    indices_xy = _gradient_resample_indices(src_x, src_y,
+                                            src_gradient_xl, src_gradient_xp,
+                                            src_gradient_yl, src_gradient_yp,
+                                            dst_x, dst_y)
+
+    if block_info:
+        y_slice, x_slice = block_info[0]["array-location"][-2:]
+        indices_xy[0, :, :] += x_slice.start
+        indices_xy[1, :, :] += y_slice.start
+
+    return indices_xy
+
+
+def block_bilinear_interpolator(src_area, dst_area, data, indices_xy, fill_value=np.nan, block_info=None, **kwargs):
+    """Bilinear interpolation implementation for resample_blocks."""
+    del src_area, dst_area
```

Comment:
Both interpolator functions don't use the src_area and dst_area?

Reply:
Indeed, but they use the previously computed and need to work with dst_area_chinks with corresponding cropped input data, so resample_locks is still needed in this case

Reply:
I guess my main question is why do these areas need to be passed if they are removed right away?

Reply:
Because `resample_blocks` tries to stay generic and has to provide these to some function that need them, for example the gradient_indices one. But now that I think about it, maybe they would fit better in the `block_info` kwarg what do you think?

Reply:
Oh that's a difficult question. I suppose we should try to imagine what a function like this would look like if it were ported to upstream xarray or dask. Those libraries wouldn't know anything about AreaDefinitions or pyresample geometry objects of course, but whatever way this geolocation or coordinate information is represented it would need to be provided. I suppose it should be `block_info`?

If `resample_blocks` was part of xarray I suppose I would expect/want a `crs` coordinate or index (something being worked on with explicit xarray indexes, currently in a PR for xarray) and `x` and `y` coordinates. So if `data` was a DataArray then they would be part of that. If this was something that dask core had, then...I'm not sure. Maybe put it in `block_info` for now.

Reply:
I put the areas in `block_info` now